homeassistant/components/ssdp/__init__.py

```diff
@@ -559,3 +594,171 @@ def _udn_from_usn(usn: str | None) -> str | None:
+                depth=24,
+                url="/static/icons/favicon-512x512.png",
+            ),
+            DeviceIcon(
+                mimetype="image/png",
+                width=384,
+                height=384,
+                depth=24,
+                url="/static/icons/favicon-384x384.png",
+            ),
+            DeviceIcon(
+                mimetype="image/png",
+                width=192,
+                height=192,
+                depth=24,
+                url="/static/icons/favicon-192x192.png",
+            ),
+        ],
+        xml=ET.Element("server_device"),
+    )
+    EMBEDDED_DEVICES: list[type[UpnpServerDevice]] = []
+    SERVICES: list[type[UpnpServerService]] = []
+
+
+async def _async_find_next_available_port(source: AddressTupleVXType) -> int:
+    """Get a free TCP port."""
+    family = socket.AF_INET if is_ipv4_address(source) else socket.AF_INET6
+    test_socket = socket.socket(family, socket.SOCK_STREAM)
+    test_socket.setblocking(False)
+    test_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+
+    for port in range(UPNP_SERVER_MIN_PORT, UPNP_SERVER_MAX_PORT):
+        try:
+            test_socket.bind(source)
+            return port
+        except OSError:
+            if port == UPNP_SERVER_MAX_PORT:
+                raise
+
+    raise RuntimeError("unreachable")
+
+
+class Server:
+    """Class to be visible via SSDP searching and advertisements."""
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize class."""
+        self.hass = hass
+        self._upnp_servers: list[UpnpServer] = []
+
+    async def async_start(self) -> None:
+        """Start the server."""
+        self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, self.async_stop)
+        await self._async_start_upnp_servers()
+
+    async def _async_get_instance_udn(self) -> str:
+        """Get Unique Device Name for this instance."""
+        instance_id = await async_get_instance_id(self.hass)
+        return f"uuid:{instance_id[0:8]}-{instance_id[8:12]}-{instance_id[12:16]}-{instance_id[16:20]}-{instance_id[20:32]}".upper()
+
+    async def _async_start_upnp_servers(self) -> None:
+        """Start the UPnP/SSDP servers."""
+        # Update UDN with our instance UDN.
+        udn = await self._async_get_instance_udn()
+        system_info = await async_get_system_info(self.hass)
+        model_name = system_info["installation_type"]
+        presentation_url = get_url(self.hass)
+        serial_number = await async_get_instance_id(self.hass)
+        HassUpnpServiceDevice.DEVICE_DEFINITION = (
+            HassUpnpServiceDevice.DEVICE_DEFINITION._replace(
+                udn=udn,
+                model_name=model_name,
+                presentation_url=presentation_url,
+                serial_number=serial_number,
+            )
+        )
+
+        # Update icon URLs.
+        base_icon_url = get_url(self.hass, allow_cloud=False, allow_external=False)
```

Comment:
When calling `get_url`, it will default to the local URL, unless the local URL has https (as that means it has an invalid cert). So I think that we can just use `presentation_url` here.